language/cc/resolve.go

```diff
@@ -49,15 +50,61 @@ func (*ccLanguage) Imports(c *config.Config, r *rule.Rule, f *rule.File) []resol
 		}
 	default:
 		hdrs := r.AttrStrings("hdrs")
+		stripIncludePrefix := r.AttrString("strip_include_prefix")
+		includePrefix := r.AttrString("include_prefix")
 		imports = make([]resolve.ImportSpec, len(hdrs))
 		for i, hdr := range hdrs {
-			imports[i] = resolve.ImportSpec{Lang: languageName, Imp: path.Join(f.Pkg, hdr)}
+			hdrRel := path.Join(f.Pkg, hdr)
+			inc := transformIncludePath(f.Pkg, stripIncludePrefix, includePrefix, hdrRel)
+			imports[i] = resolve.ImportSpec{Lang: languageName, Imp: inc}
 		}
 	}
 
 	return imports
 }
 
+// transformIncludePath converts a path to a header file into a string by which the
+// header file may be included, accounting for the library's
+// strip_include_prefix and include_prefix attributes.
+//
+// libRel is the slash-separated, repo-root-relative path to the directory
+// containing the target.
+//
+// stripIncludePrefix is the value of the target's strip_include_prefix
+// attribute. If it's "", this has no effect. If it's a relative path (including
+// "."), both libRel and stripIncludePrefix are stripped from rel. If it's an
+// absolute path, the leading '/' is removed, and only stripIncludePrefix is
+// removed from hdrRel.
+//
+// includePrefix is the value of the target's include_prefix attribute.
+// It's prepended to hdrRel after stripIncludePrefix is applied.
+//
+// hdrRel is the slash-separated, repo-root-relative path to the header file.
+func transformIncludePath(libRel, stripIncludePrefix, includePrefix, hdrRel string) string {
+	// Clean prefixes. We trust libRel and hdrRel, but the prefixes come from
+	// the build file directly.
+	if stripIncludePrefix != "" {
+		stripIncludePrefix = path.Clean(stripIncludePrefix)
+	}
+	if includePrefix != "" {
+		includePrefix = path.Clean(includePrefix)
+	}
```

Comment:
nit: shouldn't we move the cleanups outside this method? Currently it's going to be executed for each of the headers. Typically these 2 would be empty strings for gazelle managed libraries, so it might not be worthy to decrease code readability over optimizations that would mostly not kick in

Reply:
Done, fair enough. I suppose the compiler would probably not figure that out automatically.